File: core/src/net/bossdragon/system/base/collision/CollisionPhases.java
package net.bossdragon.system.base.collision;

import java.util.TreeMap;

/**
 *
 * @author Namek
 * @todo **Pleease**, optimizee meee!
 */
public class CollisionPhases {
    public static final int None = 0;
    public static final int Entered = 1;
    public static final int Existing = 2;

    /** Maps entity1Id to entity2d which maps to phase  */
    private TreeMap<Integer, TreeMap<Integer, Integer>> collisionPhases = new TreeMap<>();


    public void set(int entity1Id, int entity2Id, int phase) {
        if (entity2Id < entity1Id) {
            int tmp = entity1Id;
            entity1Id = entity2Id;
            entity2Id = tmp;
        }

        TreeMap<Integer, Integer> relations = collisionPhases.get(entity1Id);

        if (relations == null) {
            relations = new TreeMap<Integer, Integer>();
            collisionPhases.put(entity1Id, relations);
        }

        relations.put(entity2Id, phase);
    }

    public int get(int entity1Id, int entity2Id) {
        if (entity2Id < entity1Id) {
            int tmp = entity1Id;
            entity1Id = entity2Id;
            entity2Id = tmp;
        }

        TreeMap<Integer, Integer> relations = collisionPhases.get(entity1Id);
        if (relations != null) {
            Integer phase = relations.get(entity2Id);

            if (phase != null) {
                return phase;
            }
        }

        return None;
    }

    public void clear(int entityId) {
        TreeMap<Integer, Integer> relations = collisionPhases.get(entityId);
        relations.clear();

        for (TreeMap<Integer, Integer> entry : collisionPhases.values()) {
            entry.remove(entityId);
        }
    }
}

File: core/src/net/bossdragon/system/base/collision/CollisionPhases.kt
package net.bossdragon.system.base.collision

import java.util.*

/**
 * @author Namek
 * *
 * @todo **Pleease**, optimizee meee!
 */
class CollisionPhases {
    companion object Phase {
        val None = 1
        val Entered = 2
        val Existing = 3
    }

    /** Maps entity1Id to entity2d which maps to phase  */
    private val collisionPhases = TreeMap<Int, MutableMap<Int, Int>>()


    operator fun set(entity1Id: Int, entity2Id: Int, phase: Int) {
        var entity1Id = entity1Id
        var entity2Id = entity2Id
        if (entity2Id < entity1Id) {
            val tmp = entity1Id
            entity1Id = entity2Id
            entity2Id = tmp
        }

        var relations: MutableMap<Int, Int>? = collisionPhases[entity1Id]

        if (relations == null) {
            relations = TreeMap<Int, Int>()
            collisionPhases.put(entity1Id, relations)
        }

        relations.put(entity2Id, phase)
    }

    operator fun get(entity1Id: Int, entity2Id: Int): Int {
        var entity1Id = entity1Id
        var entity2Id = entity2Id
        if (entity2Id < entity1Id) {
            val tmp = entity1Id
            entity1Id = entity2Id
            entity2Id = tmp
        }

        val relations = collisionPhases[entity1Id]
        if (relations != null) {
            val phase = relations[entity2Id]

            if (phase != null) {
                return phase
            }
        }

        return Phase.None
    }

    fun clear(entityId: Int) {
        val relations = collisionPhases[entityId]

        relations?.clear()

        for (entry in collisionPhases.values) {
            entry.remove(entityId)
        }
    }
}
